spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTransactionManager.java

```diff
@@ -191,6 +191,9 @@ protected void doCommit(DefaultTransactionStatus defaultTransactionStatus)
 			if (tx.transactionManager.getState() == TransactionManager.TransactionState.STARTED) {
 
 				tx.transactionManager.commit();
+
+				// Explicitly closing read-only contexts is required because committing does not do so.
+				tx.getTransactionContext().close();
```

Comment:
@elefeint no, we always set the transaction context in `doBegin`. We set it either with real context for read-write transactions or we make that anonymous inline class for read-only.

Reply:
@ChengyuanZhao What about my question about conditional `close()`?

Reply:
@meltsufin I can make an issue for that, but my feeling is that it's not a bad thing to call close() on a readwrite context as well. 

To be honest I was surprised calling commit() on the TransactionManager didn't close the readonlyTransaction (yup ,the client lib calls it "ReadOnlyTransaction") when its commit() was called but it does seem to do so for readwrite transactions. 

I don't know if we're getting lucky in the past that close() was not necessary readwrite transactions, but I feel like it's OK to shut the door on the possibility in the future.

Reply:
Since the read-only transaction is not managed by the `TransactionManager` at all, it makes sense that `commit()` would do nothing.
That's fine as long as calling `close()` the second time is no-op, which is what it looks like.

Reply:
Ah yes you're right. We're getting the readonlyTransaction from the databaseclient instead of the transaction manager.